File: chart/python/spectralsequence_chart/chart_edge.py
from abc import ABC, abstractmethod
from .page_property import (PageProperty, PagePropertyOrValue, ensure_page_property)
from .signal_dict import SignalDict
from uuid import uuid4
from .chart_class import ChartClass, ChartClassStyle

from typing import Optional, TYPE_CHECKING, Any, Dict, cast, Union, Optional, Tuple
if TYPE_CHECKING:
    from .chart import SseqChart
    from .chart_class import ChartClass, ChartClassStyle
    from .display_primitives import Shape

from .display_primitives import UUID_str, Color, DashPattern, LineWidth, ArrowTip

class ChartEdgeStyle:
    def __init__(self, 
        action : str = "",
        color : Color = Color.BLACK,
        dash_pattern : DashPattern = [],
        line_width : float = 2,
        start_tip : Optional[ArrowTip] = None,
        end_tip : Optional[ArrowTip] = None,
    ):
        self._action : str = ""
        self._color : Color = color
        self._dash_pattern : DashPattern = dash_pattern
        self._line_width : float = line_width
        self._start_tip : Optional[ArrowTip] = start_tip
        self._end_tip : Optional[ArrowTip] = end_tip


    def to_json(self) -> Dict[str, Any]:
        return dict(
            type=type(self).__name__,
            action = self._action,
            color = self._color,
            dash_pattern = self._dash_pattern,
            line_width = self._line_width,
            start_tip = self._start_tip,
            end_tip = self._end_tip,
        )

    @classmethod
    def from_json(cls, json : Dict[str, Any]) -> "ChartEdgeStyle":
        assert json.pop("type") == cls.__name__
        return cls(**json)

    @property
    def action(self) -> str:
        """ The action of the edge. """
        return self._action

    @action.setter
    def action(self, v : str):
        self._action = v

    @property
    def color(self) -> Color:
        """ The color of the edge. """
        return self._color

    @color.setter
    def color(self, v : Color):
        self._color = v

    @property
    def dash_pattern(self) -> DashPattern:
        """The dash pattern of the edge. A dash pattern is represented as a list of positive integers."""
        return self._dash_pattern

    @dash_pattern.setter
    def dash_pattern(self, v : DashPattern):
        self._dash_pattern = v

    @property
    def line_width(self) -> LineWidth:
        """The width of the edge."""
        return self._line_width

    @line_width.setter
    def line_width(self, v : LineWidth):
        self._line_width = v

    @property
    def bend(self) -> float:
        """The bend angle of the edge. If bend is nonzero, the edge is drawn as a circular arc through the start and end points,
           where the angle between the edge from the start to the end and the tangent vector at the start point is specified by "bend".
        """
        return self._bend

    @bend.setter
    def bend(self, v : float):
        self._bend = v

    @property
    def start_tip(self) -> Optional[ArrowTip]:
        """ The start arrow tip. TODO: Explain how we represent arrow tips? """
        return self._start_tip

    @start_tip.setter
    def start_tip(self, v : Optional[ArrowTip]):
        self._start_tip = v

    @property
    def end_tip(self) -> Optional[ArrowTip]:
        """ The end arrow tip. TODO: Explain how we represent arrow tips? """
        return self._end_tip

    @end_tip.setter
    def end_tip(self, v : Optional[ArrowTip]):
        self._end_tip = v
    
    def __repr__(self):
        if self.action:
            return f"EdgeStyle({self.action})"
        d = self.to_json()
        del d["type"]
        del d["action"]
        return f"EdgeStyle({d})"


class ChartEdge(ABC):
    """ ChartEdge is the base class of ChartStructline, ChartDifferential, and ChartExtension. """
    def __init__(self, source_uuid : UUID_str, target_uuid : UUID_str):
        """ Do not call SseqEdge constructor directly, use instead SseqChart.add_structline(),
            SseqChart.add_differential(), SseqChart.add_extension(), or JSON.parse()."""
        self._sseq : SseqChart = None
        self._source_uuid = source_uuid
        self._target_uuid = target_uuid
        self._source : ChartClass
        self._target : ChartClass        
        self._uuid = str(uuid4())
        self._user_data : SignalDict[Any] = SignalDict({}, parent=self) # type: ignore

    def __repr__(self):
        fields = [repr(x) for x in [self.source, self.target]]
        return f"{type(self).__name__}({', '.join(fields)})"


    def _needs_update(self):
        if self._sseq:
            self._sseq._add_edge_to_update(self)

    def replace_source(self, style : Union[ChartClassStyle, str]) -> "ChartEdge":
        """ Calls `self.source.replace(style) <ChartClass.replace>`. 

            Arguments:
                style (ChartClassStyle | str): The style to set. If ``style`` is a string,
                    then the appropriate style is looked up in the dictionary `SseqChart.chart_class_styles`.
                    Otherwise, we use the `ChartClassStyle` provided.

            Returns:
                self (chainable)
        """
        self.source.replace(style)
        return self
    
    def replace_target(self, style : Union[ChartClassStyle, str]) -> "ChartEdge":
        """ Calls `self.target.replace(style) <ChartClass.replace>`. 
        
            Arguments:
                style (ChartClassStyle | str): The style to set. If ``style`` is a string,
                    then the appropriate style is looked up in the dictionary `SseqChart.chart_class_styles`.
                    Otherwise, we use the `ChartClassStyle` provided.

            Returns:
                self (chainable) 
        """
        self.target.replace(style)
        return self

    def delete(self):
        """ Deletes the edge. """
        self._sseq._add_edge_to_delete(self)
        del self._sseq._edges[self.uuid]
        del self.source.edges[self.source.edges.index(self)]
        del self.target.edges[self.target.edges.index(self)]

    _EDGE_TYPE_DICT : Dict[str, type]
    @staticmethod
    def from_json(json : Dict[str, Any]) -> "ChartEdge":
        if not hasattr(ChartEdge, "EDGE_TYPE_DICT"):
            ChartEdge._EDGE_TYPE_DICT = {edge_type.__name__ : edge_type for edge_type in [ChartStructline, ChartDifferential, ChartExtension]}
        edge_type = json["type"]
        if edge_type in ChartEdge._EDGE_TYPE_DICT:
            init_args = {}
            for key in ["source_uuid", "target_uuid", "page"]:
                if key in json:
                    init_args[key] = json.pop(key)
            edge = ChartEdge._EDGE_TYPE_DICT[edge_type](**init_args)
            edge._from_json_helper(**json)
            return edge
        else:
            type_names = list(ChartEdge._EDGE_TYPE_DICT.keys())
            types_list = ",".join(f'"{type}"' for type in type_names[:-1])
            types_list += f', or "${type_names[-1]}"'
            raise ValueError(f'"edge_type" should be one of {types_list}, not "{edge_type}"')

    @property
    def uuid(self) -> str:
        """ A unique id for the edge. For internal use. """
        return self._uuid

    @property
    def source(self) -> "ChartEdge":
        """ The start class of the edge. (All edges are directed.) """
        return self._source

    @property
    def target(self) -> "ChartEdge":
        """ The end class of the edge. (All edges are directed.) """
        return self._target

    def _from_json_helper(self,
        type : Optional[str], 
        uuid : UUID_str,
        user_data : Dict[str, Any],
    ):
        assert type == self.__class__.__name__
        self._uuid = uuid
        self._user_data = SignalDict(user_data, parent=self)
        

    @abstractmethod
    def to_json(self) -> Dict[str, Any]:
        return dict(
            type=self.__class__.__name__,
            uuid=self.uuid,
            source_uuid=self._source_uuid,
            target_uuid=self._target_uuid,
            user_data=self._user_data
        )

class ChartStructline(ChartEdge):
    """ A ChartStructline is an edge used to represent some sort of algebraic data on the spectral sequence. 
        Structlines are typically used to represent multiplication by a few standard elements.
        A structline will appear on page_range (<page>, <max_differential_length>) if structline.visible[<page>] 
        is true and both the source and the target class of the structure line are visible.
    """
    def __init__(self, source_uuid : UUID_str, target_uuid : UUID_str):
        super().__init__(source_uuid, target_uuid)
        self.action = ""
        self.color = (0, 0, 0, 1)
        self.dash_pattern = []
        self.line_width = 2
        self.bend = 0
        self.start_tip = None
        self.end_tip = None
        self.visible = True

    def set_style(self, style : Union[str, ChartEdgeStyle], page : Union[int, Tuple[int, int]] = None ) -> "ChartStructline":
        """ Sets the display style of the structline. 
            
            Arguments:
                style (ChartEdgeStyle | str): The style to set. If ``style`` is a string,
                    then the appropriate style is looked up in the dictionary `SseqChart.chart_edge_styles`.
                    Otherwise, we use the `ChartEdgeStyle` provided.
                
                page (int | Tuple[int, int]):
                    If argument ``page`` is omitted or ``None`` then the style is set on all pages.
                    If ``page`` is a single integer, then the stlye is set starting on that page and all later pages.
                    If ``page`` is a pair of integers, the style is set on that range of pages inclusive of the lower 
                    endpoint and exclusive of the upper endpoint.
        """ 
        if page is None:
            page = slice(None)
        if isinstance(page, (tuple, list)):
            page = slice(page[0], page[1])
        if type(style) is str:
            style = self._sseq._edge_styles[style]
        self.action[page] = style.action
        self.color[page] = style.color
        self.dash_pattern[page] = style.dash_pattern
        self.line_width[page] = style.line_width
        self.start_tip[page] = style.start_tip
        self.end_tip[page] = style.end_tip
        return self
    
    def get_style(self, page : int = 0) -> ChartEdgeStyle:
        """ Gets the display style of the structline.

            Args:
                page (int): The page on which to get the display style.
        """
        return ChartEdgeStyle(
            action = self.action,
            color = self.color[page],
            dash_pattern=self.dash_pattern[page],
            line_width=self.line_width[page],
            start_tip=self.start_tip[page],
            end_tip=self.end_tip[page]
        )
    
    def to_json(self) -> Dict[str, Any]:
        return dict(
            super().to_json(),
            type=self.__class__.__name__,
            uuid=self.uuid,
            source_uuid=self._source_uuid,
            target_uuid=self._target_uuid,
            action=self.action,
            color=self.color,
            dash_pattern=self.dash_pattern,
            line_width=self.line_width,
            bend=self.bend,
            start_tip=self.start_tip,
            end_tip=self.end_tip,
            visible=self.visible,
            user_data=self._user_data
        )

    def _from_json_helper(self,
        action : PagePropertyOrValue[str], 
        color : PagePropertyOrValue[Color],
        dash_pattern : PagePropertyOrValue[DashPattern],
        line_width : PagePropertyOrValue[float],
        bend : PagePropertyOrValue[float],
        start_tip : PagePropertyOrValue[Optional[ArrowTip]],
        end_tip : PagePropertyOrValue[Optional[ArrowTip]],
        visible : PagePropertyOrValue[bool],
        **kwargs,
    ) -> "ChartEdge":
        super()._from_json_helper(**kwargs)
        self._action =cast(PageProperty[str], ensure_page_property(action, parent=self)) 
        self._color = cast(PageProperty[Color], ensure_page_property(color, parent=self))
        self._dash_pattern = cast(PageProperty[DashPattern], ensure_page_property(dash_pattern, parent=self))
        self._line_width = cast(PageProperty[Union[float, str]], ensure_page_property(line_width, parent=self))
        self._bend = cast(PageProperty[float], ensure_page_property(bend, parent=self)) 
        self._start_tip = cast(PageProperty[Optional[ArrowTip]], ensure_page_property(start_tip, parent=self)) 
        self._end_tip = cast(PageProperty[Optional[ArrowTip]], ensure_page_property(end_tip, parent=self)) 
        self._visible = cast(PageProperty[bool], ensure_page_property(visible, parent=self))

    @property
    def action(self) -> PageProperty[str]:
        """ The action of the edge. Similar to `ChartClass.group_name`. """
        return self._action

    @action.setter
    def action(self, v : PagePropertyOrValue[str]):
        self._action = ensure_page_property(v, parent=self)
        self._needs_update()

    @property
    def color(self) -> PageProperty[Color]:
        """ The color of the edge. """
        return self._color

    @color.setter
    def color(self, v : PagePropertyOrValue[Color]): # type: ignore
        pp = ensure_page_property(v, parent=self)
        def callback():
            if self._sseq:
                pp.map_values_in_place(self._sseq.get_color)
        callback()
        pp.set_callback(callback)
        self._color = pp
        self._needs_update()

    @property
    def dash_pattern(self) -> PageProperty[DashPattern]:
        """The dash pattern of the edge. A dash pattern is represented as a list of positive integers."""
        return self._dash_pattern

    @dash_pattern.setter
    def dash_pattern(self, v : PagePropertyOrValue[DashPattern]): # type: ignore
        self._dash_pattern = ensure_page_property(v, parent=self)
        self._needs_update()

    @property
    def line_width(self) -> PageProperty[LineWidth]:
        """ The width of the edge. """
        return self._line_width


    @line_width.setter
    def line_width(self, v : PagePropertyOrValue[LineWidth]): # type: ignore
        self._line_width = ensure_page_property(v, parent=self)
        self._needs_update()

    @property
    def bend(self) -> PageProperty[float]:
        """ The bend angle of the edge. If bend is nonzero, the edge is drawn as a circular arc through the start and end points,
            where the angle between the edge from the start to the end and the tangent vector at the start point is specified by "bend".
        """
        return self._bend

    @bend.setter
    def bend(self, v : PagePropertyOrValue[float]): # type: ignore
        self._bend = ensure_page_property(v, parent=self)
        self._needs_update()

    @property
    def visible(self) -> PageProperty[bool]:
        """ Is the structline visible on the given page? """
        return self._visible

    @visible.setter
    def visible(self, v : PagePropertyOrValue[bool]): # type: ignore
        self._visible = ensure_page_property(v, parent=self)
        self._needs_update()

    @property
    def start_tip(self) -> PageProperty[Optional[ArrowTip]]:
        """ The start arrow tip. TODO: Explain how we represent arrow tips """
        return self._start_tip


    @start_tip.setter
    def start_tip(self, v : PagePropertyOrValue[Optional[ArrowTip]]):
        self._start_tip = ensure_page_property(v, parent=self)
        self._needs_update()

    @property
    def end_tip(self) -> PageProperty[Optional[ArrowTip]]:
        """ The end arrow tip. TODO: Explain how we represent arrow tips """
        return self._end_tip

    @end_tip.setter
    def end_tip(self, v : PagePropertyOrValue[Optional[ArrowTip]]):
        self._end_tip = ensure_page_property(v, parent=self)
        self._needs_update()

class SinglePageChartEdge(ChartEdge):
    """ SinglePageChartEdge handles most of the common code between ChartDifferential and ChartExtension. """
    def __init__(self, source_uuid : UUID_str, target_uuid : UUID_str):
        super().__init__(source_uuid, target_uuid)
        self._action = ""
        self._color : Color = (0, 0, 0, 1)
        self._dash_pattern = []
        self._line_width = 3
        self._bend = 0
        self._start_tip = None
        self._end_tip = None
        self._visible = True

    def set_style(self, style : ChartEdgeStyle) -> "SinglePageChartEdge":
        """ Sets the display style of the edge. 
            
            Arguments:
                style (ChartEdgeStyle | str): The style to set. If ``style`` is a string,
                    then the appropriate style is looked up in the dictionary `SseqChart.chart_edge_styles`.
                    Otherwise, we use the `ChartEdgeStyle` provided.
            
            Returns:
                self (chainable)
        """         
        self.action = style.action
        self.color = style.color
        self.dash_pattern = style.dash_pattern
        self.line_width = style.line_width
        self.start_tip = style.start_tip
        self.end_tip = style.end_tip
        return self
    
    def get_style(self) -> ChartEdgeStyle:
        """ Get the style of the edge """ 
        return ChartEdgeStyle(
            action = self.action,
            color = self.color,
            dash_pattern=self.dash_pattern,
            line_width=self.line_width,
            start_tip=self.start_tip,
            end_tip=self.end_tip
        )

    def to_json(self) -> Dict[str, Any]:
        return dict(
            super().to_json(),
            type=self.__class__.__name__,
            uuid=self.uuid,
            source_uuid=self._source_uuid,
            target_uuid=self._target_uuid,
            action=self.action,
            color=self.color,
            start_tip=self.start_tip,
            end_tip=self.end_tip,
            dash_pattern=self.dash_pattern,
            line_width=self.line_width,
            bend=self.bend,
            visible=self.visible,
            user_data=self._user_data
        )

    def _from_json_helper(self,
        action : str,
        color : Color,
        dash_pattern : DashPattern,
        line_width : float,
        bend : float,
        start_tip : Optional[ArrowTip],
        end_tip : Optional[ArrowTip],
        visible : bool,
        **kwargs,
    ) -> "ChartEdge":
        super()._from_json_helper(**kwargs)
        self._action = action
        self._color = color
        self._dash_pattern = dash_pattern
        self._line_width = line_width
        self._bend = bend
        self._start_tip = start_tip
        self._end_tip = end_tip
        self._visible = visible

    @property
    def action(self) -> str:
        """ The action of the edge. """
        return self._action

    @action.setter
    def action(self, v : str):
        self._action = v
        self._needs_update()

    @property
    def color(self) -> Color:
        """ The color of the edge. """
        return self._color

    @color.setter
    def color(self, v : Color):
        if self._sseq:
            self._color = self._sseq.get_color(v)
        else:
            self._color = v
        self._needs_update()

    @property
    def dash_pattern(self) -> DashPattern:
        """ The dash pattern of the edge. A dash pattern is represented as a list of positive integers. """
        return self._dash_pattern

    @dash_pattern.setter
    def dash_pattern(self, v : DashPattern):
        self._dash_pattern = v
        self._needs_update()

    @property
    def line_width(self) -> LineWidth:
        """ The width of the edge. """
        return self._line_width


    @line_width.setter
    def line_width(self, v : LineWidth):
        self._line_width = v
        self._needs_update()

    @property
    def bend(self) -> float:
        """ The bend angle of the edge. If bend is nonzero, the edge is drawn as a circular arc through the start and end points,
            where the angle between the edge from the start to the end and the tangent vector at the start point is specified by "bend".
        """
        return self._bend

    @bend.setter
    def bend(self, v : float):
        self._bend = v
        self._needs_update()

    @property
    def visible(self) -> bool:
        """ Is the structline visible on the given page? """
        return self._visible

    @visible.setter
    def visible(self, v : bool):
        self._visible = v
        self._needs_update()


    @property
    def start_tip(self) -> Optional[ArrowTip]:
        """ The start arrow tip. TODO: Explain how we represent arrow tips. """
        return self._start_tip


    @start_tip.setter
    def start_tip(self, v : Optional[ArrowTip]):
        self._start_tip = v
        self._needs_update()

    @property
    def end_tip(self) -> Optional[ArrowTip]:
        """ The end arrow tip. TODO: Explain how we represent arrow tips. """
        return self._end_tip

    @end_tip.setter
    def end_tip(self, v : Optional[ArrowTip]):
        self._end_tip = v
        self._needs_update()


class ChartDifferential(SinglePageChartEdge):
    """ A ChartDifferential is an edge used to represent the behavior of the differential on the spectral sequence. 
        A chart differential will appear on page_range (<page>, <max_differential_length>) if <page> <= differential.page <= <max_differential_length>
        and if both the source and target of the differential appear on page <page>.
    """
    def __init__(self, source_uuid : UUID_str, target_uuid : UUID_str, page : int):
        super().__init__(source_uuid, target_uuid)
        self.page : int = page

    def to_json(self) -> Dict[str, Any]:
        return dict(
            super().to_json(),
            page=self.page
        )

class ChartExtension(SinglePageChartEdge):
    """ A ChartExtension is an edge used to represent extensions in the spectral sequence. 
        Generally extensions represent the same sort of algebraic structure as the structlines.
        A chart extension will appear on page_range (<page>, <max_differential_length>) if <page> == infinity
        and both the source and the target of the extension appear on page infinity.
    """
    pass

File: chart/chart/javascript/src/ChartEdge.ts
import { ChartClass } from './ChartClass';
import {
    PageProperty,
    PagePropertyOrValue,
    initialPagePropertyValue,
    initialOptionalPagePropertyValue,
    pagePropertyOrValueToPageProperty,
    PageProperties,
    PagePropertyOrValues,
} from './PageProperty';
import { INFINITY } from './infinity';
import { v4 as uuidv4 } from 'uuid';
import { Color, DashPattern, Black } from './Color';
import { SseqChart, MessageUpdate } from './SseqChart';
import { Walker } from './json_utils';

const DefaultLineWidth: number = 3;

export class ArrowTip {
    tip: string;
    type: string = "ArrowTip";
    constructor({tip, type } : {tip? : string, type: "ArrowTip"}){
        this.tip = tip || "None";
    }
    toJSON(){
        if(this.tip === "None"){
            return null;
        } else {
            return this;
        }
    }
    static fromJSON(walker : Walker, obj : any){
        return new ArrowTip(obj);
    }
}
const NoTip = new ArrowTip({type : "ArrowTip"});

export interface EdgeStyle {
    start_tip: ArrowTip;
    end_tip: ArrowTip;
    bend: number;
    color: Color;
    dash_pattern: DashPattern;
    line_width: number;
    visible: boolean;
    action : string;
}

export interface ChartEdgeConstructorArgs {
    source_uuid?: string;
    target_uuid?: string;
    type?: string;
    uuid?: string;
    user_data?: Map<string, any>;
}

export interface ChartStructlineConstructorArgs
    extends ChartEdgeConstructorArgs,
        PagePropertyOrValues<EdgeStyle> {
    type?: 'ChartStructline';
}

export interface ChartDifferentialConstructorArgs
    extends ChartEdgeConstructorArgs,
        EdgeStyle {
    page: number;
    type?: 'ChartDifferential';
}

export interface ChartExtensionConstructorArgs
    extends ChartEdgeConstructorArgs,
        EdgeStyle {
    type?: 'ChartExtension';
}

export abstract class ChartEdge {
    _sseq?: SseqChart;
    uuid: string;
    source_uuid: string;
    target_uuid: string;
    source?: ChartClass;
    target?: ChartClass;
    user_data: Map<string, any>;

    constructor(kwargs: ChartEdgeConstructorArgs) {
        if (!kwargs.source_uuid) {
            throw TypeError(`Missing mandatory argument "source_uuid"`);
        }
        if (!kwargs.target_uuid) {
            throw TypeError(`Missing mandatory argument "target_uuid"`);
        }
        let errorContext = '';
        this.uuid = kwargs.uuid || uuidv4();

        this.source_uuid = kwargs.source_uuid;
        this.target_uuid = kwargs.target_uuid;
        this.user_data = kwargs.user_data || new Map();
    }

    abstract update(kwargs: MessageUpdate<ChartEdge>): void;

    protected _baseUpdate(msg: MessageUpdate<ChartEdge>) {
        let kwargs = msg.update_fields;
        // if(kwargs.source_uuid){
        //     if(this._source_uuid !== kwargs.source_uuid){
        //         throw TypeError(`Inconsistent values for "source_uuid".`);
        //     }
        // }
        // if(kwargs.target_uuid){
        //     if(this._target_uuid !== kwargs.target_uuid){
        //         throw TypeError(`Inconsistent values for "target_uuid".`);
        //     }
        // }
        if (kwargs.uuid) {
            if (this.uuid !== kwargs.uuid) {
                throw TypeError(`Inconsistent values for "uuid".`);
            }
        }
        // if(kwargs.type){
        //     if(kwargs.type !== this.constructor.name){
        //         throw TypeError(`Invalid value for "type"`)
        //     }
        // }
        this.user_data = this.user_data || new Map();
    }

    abstract getEdgeStyle(page: number): EdgeStyle;

    delete() {
        this.source?.edges.delete(this);
        this.target?.edges.delete(this);
        this._sseq!.edges.delete(this.uuid);
    }

    abstract drawOnPageQ(pageRange: [number, number]): boolean;

    toJSON(): any {
        return {
            type: this.constructor.name,
            uuid: this.uuid,
            source_uuid: this.source_uuid,
            target_uuid: this.target_uuid,
        };
    }
}

export class ChartStructline extends ChartEdge {
    start_tip: PageProperty<ArrowTip>;
    end_tip: PageProperty<ArrowTip>;
    bend: PageProperty<number>;
    color: PageProperty<Color>;
    dash_pattern: PageProperty<DashPattern>;
    line_width: PageProperty<number>;
    visible: PageProperty<boolean>;
    action: PageProperty<string>;
    constructor(kwargs: ChartStructlineConstructorArgs) {
        super(kwargs);
        let errorContext = '';
        this.visible = initialPagePropertyValue(
            kwargs.visible,
            true,
            'visible',
            errorContext,
        );
        this.start_tip = initialPagePropertyValue(
            kwargs.start_tip,
            NoTip,
            'start_tip',
            errorContext,
        );
        this.end_tip = initialPagePropertyValue(
            kwargs.end_tip,
            NoTip,
            'end_tip',
            errorContext,
        );
        this.bend = initialPagePropertyValue(
            kwargs.bend,
            0,
            'bend',
            errorContext,
        );
        this.color = initialPagePropertyValue(
            kwargs.color,
            Black,
            'color',
            errorContext,
        );
        this.dash_pattern = initialPagePropertyValue(
            kwargs.dash_pattern,
            [],
            'dash_pattern',
            errorContext,
        );
        this.line_width = initialPagePropertyValue(
            kwargs.line_width,
            3,
            'line_width',
            errorContext,
        );
        this.action = initialPagePropertyValue(
            kwargs.action,
            "",
            'action',
            errorContext,
        );
    }

    update(msg: MessageUpdate<ChartStructline>) {
        super._baseUpdate(msg);
        let kwargs = msg.update_fields;
        if (kwargs.visible) {
            this.visible = pagePropertyOrValueToPageProperty(kwargs.visible);
        }

        if (kwargs.start_tip) {
            this.start_tip = pagePropertyOrValueToPageProperty(
                kwargs.start_tip,
            );
        }

        if (kwargs.end_tip) {
            this.end_tip = pagePropertyOrValueToPageProperty(kwargs.end_tip);
        }

        if (kwargs.color) {
            this.color = pagePropertyOrValueToPageProperty(kwargs.color);
        }
        if (kwargs.dash_pattern) {
            this.dash_pattern = pagePropertyOrValueToPageProperty(
                kwargs.dash_pattern,
            );
        }
        if (kwargs.line_width) {
            this.line_width = pagePropertyOrValueToPageProperty(
                kwargs.line_width,
            );
        }
        if (kwargs.bend) {
            this.bend = pagePropertyOrValueToPageProperty(kwargs.bend);
        }
    }

    drawOnPageQ(pageRange: [number, number]): boolean {
        return (
            this.visible[pageRange[0]] &&
            this.source!.drawOnPageQ(pageRange[0]) &&
            this.target!.drawOnPageQ(pageRange[0])
        );
    }

    getEdgeStyle(page: number): EdgeStyle {
        return {
            start_tip: this.start_tip[page],
            end_tip: this.end_tip[page],
            bend: this.bend[page],
            color: this.color[page],
            dash_pattern: this.dash_pattern[page],
            line_width: this.line_width[page],
            visible: this.visible[page],
            action: this.action[page],
        };
    }

    toJSON(): any {
        let result = super.toJSON();
        let edge_style: PageProperties<EdgeStyle> & { user_data: any, action : any } = {
            visible: this.visible,
            color: this.color,
            dash_pattern: this.dash_pattern,
            line_width: this.line_width,
            bend: this.bend,
            start_tip: this.start_tip,
            end_tip: this.end_tip,
            user_data: this.user_data,
            action : this.action,
        };
        Object.assign(result, edge_style);
        return result;
    }

    static fromJSON(walker: Walker, obj: any): ChartStructline {
        return new ChartStructline(obj);
    }
}

abstract class SinglePageChartEdge extends ChartEdge {
    start_tip?: ArrowTip;
    end_tip?: ArrowTip;
    bend: number;
    color: Color;
    dash_pattern: DashPattern;
    line_width: number;
    visible: boolean;
    action: string;
    constructor(kwargs: ChartEdgeConstructorArgs & Partial<EdgeStyle>) {
        super(kwargs);
        this.start_tip = kwargs.start_tip;
        this.end_tip = kwargs.end_tip;
        this.bend = kwargs.bend !== undefined ? kwargs.bend : 0;
        this.color = kwargs.color || Black;
        this.dash_pattern = kwargs.dash_pattern || [];
        this.line_width = kwargs.line_width !== undefined ? kwargs.line_width : DefaultLineWidth;
        this.visible = kwargs.visible !== undefined ? kwargs.visible : true;
        this.action = kwargs.action || "";
    }

    _baseUpdate(msg: MessageUpdate<SinglePageChartEdge>) {
        let kwargs = msg.update_fields;
        if (kwargs.visible) {
            this.visible = kwargs.visible;
        }
        if (kwargs.color) {
            this.color = kwargs.color;
        }
        if (kwargs.dash_pattern) {
            this.dash_pattern = kwargs.dash_pattern;
        }
        if (kwargs.line_width) {
            this.line_width = kwargs.line_width;
        }
        if (kwargs.bend) {
            this.bend = kwargs.bend;
        }
        if (kwargs.start_tip) {
            this.start_tip = kwargs.start_tip;
        }

        if (kwargs.end_tip) {
            this.end_tip = kwargs.end_tip;
        }
    }

    getEdgeStyle(_page: number): EdgeStyle {
        return {
            start_tip: this.start_tip || NoTip,
            end_tip: this.end_tip || NoTip,
            bend: this.bend,
            color: this.color,
            dash_pattern: this.dash_pattern,
            line_width: this.line_width,
            visible: this.visible,
            action : this.action,
        };
    }

    toJSON(): any {
        let result = super.toJSON();
        Object.assign(result, this.getEdgeStyle(0));
        result.user_data = this.user_data;
        return result;
    }
}

export class ChartDifferential extends SinglePageChartEdge {
    page: number;
    constructor(kwargs: ChartDifferentialConstructorArgs) {
        super(kwargs);
        this.page = kwargs.page;
    }

    update(kwargs: MessageUpdate<ChartDifferential>) {
        super._baseUpdate(kwargs);
    }

    drawOnPageQ(pageRange: [number, number]) {
        return (
            pageRange[0] === 0 ||
            (pageRange[0] <= this.page && this.page <= pageRange[1])
        );
    }

    toJSON(): any {
        let result = super.toJSON();
        result.page = this.page;
        return result;
    }

    static fromJSON(walker: Walker, obj: any): ChartDifferential {
        return new ChartDifferential(obj);
    }
}

export class ChartExtension extends SinglePageChartEdge {
    constructor(kwargs: ChartExtensionConstructorArgs) {
        super(kwargs);
    }

    update(kwargs: MessageUpdate<ChartExtension>) {
        super._baseUpdate(kwargs);
    }

    drawOnPageQ(pageRange: [number, number]): boolean {
        return pageRange[0] === INFINITY;
    }

    static fromJSON(walker: Walker, obj: any): ChartExtension {
        return new ChartExtension(obj);
    }
}
